cheetah/bmadx_utils.py

```diff
@@ -0,0 +1,148 @@
+import torch
+
+
+def cheetah_to_bmad_coords(
+    cheetah_coords: torch.Tensor, ref_energy: torch.Tensor, mc2: torch.Tensor
+) -> torch.Tensor:
+    """Transforms Cheetah coordinates to Bmad coordinates.
+    :param cheetah_coords: 7-dimensional particle vectors in Cheetah coordinates.
+    :param ref_energy: reference energy in eV.
+    """
+    # initialize Bmad coordinates:
+    bmad_coords = cheetah_coords[..., :6].clone()
+
+    # Cheetah longitudinal coords:
+    tau = cheetah_coords[..., 4]
+    delta = cheetah_coords[..., 5]
+
+    # compute p0c and bmad z, pz
+    p0c = torch.sqrt(ref_energy**2 - mc2**2)
+    energy = ref_energy + delta * p0c
+    p = torch.sqrt(energy**2 - mc2**2)
+    beta = p / energy
+    z = -beta * tau
+    pz = (p - p0c) / p0c
+
+    # Bmad coords:
+    bmad_coords[..., 4] = z
+    bmad_coords[..., 5] = pz
+
+    return bmad_coords, p0c
+
+
+def bmad_to_cheetah_coords(
+    bmad_coords: torch.Tensor, p0c: torch.Tensor, mc2: torch.Tensor
+) -> torch.Tensor:
+    """Transforms Bmad coordinates to Cheetah coordinates.
+    :param bmad_coords: 6-dimensional particle vectors in Bmad coordinates.
+    :param p0c: reference momentum in eV/c.
+    """
+    # initialize Cheetah coordinates:
+    cheetah_coords = torch.ones(
+        (*bmad_coords.shape[:-1], 7), dtype=bmad_coords.dtype, device=bmad_coords.device
+    )
+    cheetah_coords[..., :6] = bmad_coords.clone()
+
+    # Bmad longitudinal coords:
+    z = bmad_coords[..., 4]
+    pz = bmad_coords[..., 5]
+
+    # compute ref_energy and Cheetah tau, delta
+    ref_energy = torch.sqrt(p0c**2 + mc2**2)
+    p = (1 + pz) * p0c
+    energy = torch.sqrt(p**2 + mc2**2)
+    beta = p / energy
+    tau = -z / beta
+    delta = (energy - ref_energy) / p0c
+
+    # Cheetah coords:
+    cheetah_coords[..., 4] = tau
+    cheetah_coords[..., 5] = delta
+
+    return cheetah_coords, ref_energy
+
+
+def offset_particle_set(x_offset, y_offset, tilt, x_lab, px_lab, y_lab, py_lab):
```

Comment:
Here is missing some docstring and variable typing

Reply:
done